File: algorithms_princeton/src/union-find/unionFind.py
#=============================================================================
#
# Author: liangrui.hlr email:i@nyankosama.com
#
# Last modified:	2014-03-11 01:42
#
# Filename:		unionFind.py
#
# Description: UnionFind operation data structure
#
#=============================================================================*/

class UnionFind:
    def __init__(self, num):
        self.idArray = range(num)
        self.size = [1] * num
    
    def union(self, node1, node2):
        #use weighted union-find
        root1 = self.__root(node1)
        root2 = self.__root(node2)
        if root1 == root2:
            return
        if self.size[root1] < self.size[root2]:
            self.idArray[root1] = root2
            self.size[root2] += self.size[root1]
        else:
            self.idArray[root2] = root1
            self.size[root1] += self.size[root2]

    def connected(self, node1, node2):
        if self.__root(node1) == self.__root(node2):
            return True
        else:
            return False

    #find the root of a given node
    def __root(self, node):
        while self.idArray[node] is not node:
            #use path compression
            self.idArray[node] = self.idArray[self.idArray[node]]
            node = self.idArray[node]
        return node


if __name__=="__main__":
    #test cases
    uf = UnionFind(10)
    uf.union(0,5)
    uf.union(5,6)
    uf.union(6,1)
    uf.union(1,2)
    uf.union(2,7)

    uf.union(8,3)
    uf.union(3,4)
    uf.union(4,9)

    print uf.connected(0,5)#True
    print uf.connected(0,7)#True
    print uf.connected(1,2)#True
    print uf.connected(5,8)#False

    print uf.connected(8,9)#True
    print uf.connected(2,9)#False

File: algorithms_princeton/union-find/cpp_version/unionFind.cpp
#include "unionFind.h"

template<int NODE_NUM>
UnionFind<NODE_NUM>::UnionFind(){
    for (int i = 0; i < NODE_NUM; i++) {
        _node_array[i] = i;
        _root_size[i] = 1;
    }
}

template<int NODE_NUM>
void UnionFind<NODE_NUM>::createUnion(int node1, int node2){
    int root1 = _root(node1);
    int root2 = _root(node2);
    if (root1 == root2)
        return;

    if (_root_size[root1] < _root_size[root2]){
        _node_array[root1] = root2;
        _root_size[root2] += _root_size[root1];
    }
    else {
        _node_array[root2] = root1;
        _root_size[root1] += _root_size[root2];
    }
}

template<int NODE_NUM>
bool UnionFind<NODE_NUM>::connected(int node1, int node2){
    if (_root(node1) == _root(node2))
        return true;
    else
        return false;
}

template<int NODE_NUM>
int UnionFind<NODE_NUM>::_root(int node){
    while (_node_array[node] != node){
        _node_array[node] = _node_array[_node_array[node]];
        node = _node_array[node];
    }
    return node;
}
